pal/inc/pal_char16.h

```diff
@@ -28,26 +28,22 @@ This file is used to define the wchar_t type as a 16-bit type on Unix.
 // vc++, for whom wchar_t is already a typedef instead of a built-in.
 
 #if !defined(PAL_STDCPP_COMPAT) || defined(USING_PAL_STDLIB)
-#if defined (PLATFORM_UNIX) && defined(__GNUC__)
-#undef wchar_t
-#define wchar_t __wchar_16_cpp__
-#endif // PLATFORM_UNIX
-
 // Set up the wchar_t type (which got preprocessed to __wchar_16_cpp__).
 // In C++11, the standard gives us char16_t, which is what we want (and matches types with u"")
 // In C, this doesn't exist, so use unsigned short.
 
 #if !defined(_WCHAR_T_DEFINED) || !defined(_MSC_VER)
 #if defined (PLATFORM_UNIX)
 #if defined(__cplusplus)
+#undef wchar_t
+#define wchar_t __wchar_16_cpp__
 typedef char16_t wchar_t;
 #else
-typedef unsigned short wchar_t;
+typedef unsigned short char16_t;
```

Comment:
@jianchun  I will try to answer all related questions here;

We did redefine `wchar_t` to `2 bytes` or however the char16_t is defined. Besides, ensured that for the rest of the library that we use `char16_t`

The problem is that unix has wchar is 32 bits, while char16_t is 2 bytes. As a result, we were redefining that 4bytes wchar back to char16_t. Somehow in the logic we were doing it before including `type_traits`. So, as a result, we were ending up redefining the same interface again and again.

IMO this PR fixes the inconsistencies around wchar_t. So we do expect sizeof(wchar_t) == sizeof(char16_t)